src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -643,9 +643,19 @@ modify = undefined
 -- | Swap the elements at two indices.
 --
 -- /Pre-release/
-swapIndices :: -- (MonadIO m, Storable a) =>
-    Array a -> Int -> Int -> m ()
-swapIndices = undefined
+swapIndices :: forall m a. (MonadIO m, Storable a)
+    => Array a -> Int -> Int -> m ()
+swapIndices Array{..} i1 i2 = liftIO $ do
+    unsafeWithArrayContents arrContents arrStart $ \ptr -> do
+        let elemSize = sizeOf (undefined :: a)
+            elemPtr1 = ptr `plusPtr` (elemSize * i1)
+            elemPtr2 = ptr `plusPtr` (elemSize * i2)
+        when (i1 < 0 ) $ invalidIndex "swapIndices" i1
+        when (i2 < 0 ) $ invalidIndex "swapIndices" i2
```

Comment:
1. We need to check if we are writing beyond the end of the array
2. we can implement `unsafeSwapIndices` to swap without any index checks and use that here. It can be useful in perf critical higher level combinators that need to swap indices.